demos/http/http_demo_s3_upload/http_demo_s3_upload.c

```diff
@@ -0,0 +1,626 @@
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/* Standard includes. */
+#include <assert.h>
+#include <stdlib.h>
+#include <string.h>
+
+/* POSIX includes. */
+#include <unistd.h>
+
+/* Include Demo Config as the first non-system header. */
+#include "demo_config.h"
+
+/* Common HTTP demo utilities. */
+#include "http_demo_utils.h"
+
+/* HTTP API header. */
+#include "core_http_client.h"
+
+/* OpenSSL transport header. */
+#include "openssl_posix.h"
+
+/* Retry utilities. */
+#include "retry_utils.h"
+
+/* Check that TLS port of the server is defined. */
+#ifndef HTTPS_PORT
+    #error "Please define a HTTPS_PORT."
+#endif
+
+/* Check that a path for Root CA Certificate is defined. */
+#ifndef ROOT_CA_CERT_PATH
+    #error "Please define a ROOT_CA_CERT_PATH."
+#endif
+
+/* Check that the pre-signed PUT URL is defined. */
+#ifndef S3_PRESIGNED_PUT_URL
+    #error "Please define a S3_PRESIGNED_PUT_URL."
+#endif
+
+/* Check that the pre-signed GET URL is defined. */
+#ifndef S3_PRESIGNED_GET_URL
+    #error "Please define a S3_PRESIGNED_GET_URL."
+#endif
+
+
+/**
+ * @brief ALPN protocol name to be sent as part of the ClientHello message.
+ *
+ * @note When using ALPN, port 443 must be used to connect to AWS IoT Core.
+ */
+#define IOT_CORE_ALPN_PROTOCOL_NAME               "\x0ex-amzn-http-ca"
+
+/**
+ * @brief Field name of the HTTP Range header to read from server response.
+ */
+#define HTTP_CONTENT_RANGE_HEADER_FIELD           "Content-Range"
+
+/**
+ * @brief Length of the HTTP Range header field.
+ */
+#define HTTP_CONTENT_RANGE_HEADER_FIELD_LENGTH    ( sizeof( HTTP_CONTENT_RANGE_HEADER_FIELD ) - 1 )
+
+/**
+ * @brief Delay in seconds between each iteration of the demo.
+ */
+#define DEMO_LOOP_DELAY_SECONDS                   ( 5U )
+
+/* Check that transport timeout for transport send and receive is defined. */
+#ifndef TRANSPORT_SEND_RECV_TIMEOUT_MS
+    #define TRANSPORT_SEND_RECV_TIMEOUT_MS    ( 1000 )
+#endif
+
+/* Check that size of the user buffer is defined. */
+#ifndef USER_BUFFER_LENGTH
+    #define USER_BUFFER_LENGTH    ( 2048 )
+#endif
+
+/* Pointer to the data to upload.*/
+#ifndef DEMO_HTTP_UPLOAD_DATA
+    #define DEMO_HTTP_UPLOAD_DATA    "Hello World!"
+#endif
+
+/* The size of the data in bytes to upload. */
+#ifndef DEMO_HTTP_UPLOAD_DATA_SIZE
+    #define DEMO_HTTP_UPLOAD_DATA_SIZE    ( sizeof( DEMO_HTTP_UPLOAD_DATA ) - 1 )
+#endif
```

Comment:
hmm why do we need to wrap these in #ifndef?

Reply:
Please change to length